### PR TITLE
[SPARK-15862] [SQL] Better Error Message When Having Database Name in CACHE TABLE AS SELECT

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

~~If the temp table already exists, we should not silently replace it when doing `CACHE TABLE AS SELECT`. This is inconsistent with the behavior of `CREAT VIEW` or `CREATE TABLE`. This PR is to fix this silent drop.~~

~~Maybe, we also can introduce new syntax for replacing the existing one. For example, in Hive, to replace a view, the syntax should be like `ALTER VIEW AS SELECT` or `CREATE OR REPLACE VIEW AS SELECT`~~

The table name in `CACHE TABLE AS SELECT` should NOT contain database prefix like "database.table". Thus, this PR captures this in Parser and outputs a better error message, instead of reporting the view already exists.

In addition, refactoring the `Parser` to generate table identifiers instead of returning the table name string.
#### How was this patch tested?
- Added a test case for caching and uncaching qualified table names
- Fixed a few test cases that do not drop temp table at the end
- Added the related test case for the issue resolved in this PR

### CODE REVIEW

**Comment by gatorsmile** [comment]
@hvanhovell @liancheng Sorry, I realized we can perfectly support caching the fully qualified table. In this PR, I just added a test case to improve the test case coverage. Thanks!

**Comment by hvanhovell** [comment]
LGTM

@liancheng anything to add?

**Comment by rxin** [comment]
Are we breaking past behavior here? I would say we shouldn't change this, since this is a weird command anyway. I'm actually surprised we have this command.

**Comment by gatorsmile** [comment]
Yeah, in Spark 1.6, we also silently drop the temporary table if the names are the same. Let me remove the related changes and update the title and JIRA

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala [hunk 1]
```diff
@@ -17,30 +17,30 @@
...
 import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
 
 case class CacheTableCommand(
-  tableName: String,
-  plan: Option[LogicalPlan],
-  isLazy: Boolean)
-  extends RunnableCommand {
+    tableIdent: TableIdentifier,
+    plan: Option[LogicalPlan],
+    isLazy: Boolean) extends RunnableCommand {
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala]: IIUC, we can add an assertion here saying that when `plan` is defined, `tableIdent.database` must be either `None` or `Some("default")`, right?

**Comment by liancheng** [comment]
Mostly LGTM except for one minor comment.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala]: Sure, will do it. Thanks!

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala]: Just added. : ) Please check if the checking is enough. https://github.com/apache/spark/pull/13572/files#diff-bc55b5f76add105ec32ae4107075b278R30

`default`.`tab` is still not allowed to create temp tables. Thus, I did not change that part. Let me know if anything else I need to change. Thanks!

**Comment by gatorsmile** [comment]
@hvanhovell @liancheng @rxin How about the latest changes? Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala (hunk 1 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala]: Yea, I think that's enough. Just one thing, usually we write SQL keywords all in capital letters.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala]: @liancheng Done. : )

**Pull request closed without merging**